### ISSUE TITLE
detect when the syslog preload hook has failed, and treat it as if it weren't there

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cassandane#85]
From mailing list, on ~RHEL7~ RHEL8, our custom syslog intercept that we LD_PRELOAD into Cyrus, seems to be silently ignored.  Nobody has any idea why yet.  When our tests go looking for syslog errors at the end, they find the expected output file nonexistent.

We might be able to work around this by the following:
* add a test binary that simply syslogs a magic word (maybe the test id, supplied as a command-line parameter)
* during the syslog setup (while we're figuring out whether or not we `have_syslog_replacement`), call this binary with the same LD_PRELOAD we would use for Cyrus
* then check for the existence of the syslog file, and for it containing the magic word, and if either are false, then we do not `have_syslog_replacement` after all
* if we don't `have_syslog_replacement`, disable the syslog components as usual

It would be very nice to understand what's really wrong (so that RHEL7 users can run Cassandane thoroughly, instead of needing to skip stuff), but in the meantime this might make it at least be able to run the non-syslog tests without a lot of distracting errors

**Comment by falon** [comment]
This is good.
In my case, the issue with LD_PRELOAD is in RHEL8. I didn't try what happens in RHEL7.

**Comment by elliefm** [comment]
This is done as of 99460e9

**Comment by falon** [comment]
I'm still experiencing problems...
Just for instance:
```
Cyrus::Flags.max_userflags                                              [FAILED]

Time: 10 wallclock secs ( 0.03 usr  0.01 sys +  6.65 cusr  0.30 csys =  6.99 CPU)

!!!FAILURES!!!
Test Results:
Run: 1, Failures: 1, Errors: 0

There was 1 failure:
1) Flags.max_userflags
  didn't match /(?^:IOERROR: out of flags)/ at Cassandane/Cyrus/Flags.pm line 478.
        Cassandane::Cyrus::Flags::test_max_userflags(Cassandane::Cyrus::Flags=HASH(0x55baeb447418)) called at /usr/share/perl5/vendor_perl/Test/Unit/TestCase.pm line 75
        [...framework calls elided...]

Annotations:
=====> Cyrus::TestCase::set_up[61] ---------- BEGIN test_max_userflags ----------
=====> Instance::start[654] start main instance for test test_max_userflags: basedir /root/rpmbuild/BUILD/cyrus-imapd-3.2.1/cassandane/work/1159120101
2020/06/04-13:59:12 Cassandane::Net::SMTPServer (type Net::Server::PreForkSimple) starting! pid(5384)
Resolved [localhost]:19100 to [::1]:19100, IPv6
Resolved [localhost]:19100 to [127.0.0.1]:19100, IPv4
Binding to TCP port 19100 on host ::1 with IPv6
Binding to TCP port 19100 on host 127.0.0.1 with IPv4
Group Not Defined.  Defaulting to EGID '12 0'
User Not Defined.  Defaulting to EUID '76'
=====> Instance::_start_smtpd[1108] started smtpd as 5384
=====> Instance::_find_binary[529] Found binary ctl_cyrusdb in /root/rpmbuild/BUILDROOT/cyrus-imapd-3.2.1-0.el8.x86_64/usr/sbin
=====> Instance::_find_binary[529] Found binary imapd in /root/rpmbuild/BUILDROOT/cyrus-imapd-3.2.1-0.el8.x86_64/usr/libexec/cyrus-imapd
=====> Instance::_fork_command[1546] Running: "/root/rpmbuild/BUILD/cyrus-imapd-3.2.1/cassandane-68e4890346edeb68463d20f394a37f0862bef2ae/utils/syslog_probe" "1159120101"
=====> Instance::_fork_command[1546] child pid=5390
=====> Instance::_fork_command[1546] setting core size limit to 104857600
=====> Instance::setup_syslog_replacement[1141] didn't find the magic word when probing syslog
=====> Instance::setup_syslog_replacement[1141] tests will not examine syslog output
=====> Instance::_start_notifyd[1142] started notifyd for /root/rpmbuild/BUILD/cyrus-imapd-3.2.1/cassandane/work/1159120101 as 5391
```

ok, it didn't find _magic_ word and `getsyslog` returns empty, but Flags.pm wants to find IOERROR into @lines. Uhm...

```
    # We should have generated an IOERROR
    my @lines = $self->{instance}->getsyslog();
    $self->assert_matches(qr/IOERROR: out of flags/, "@lines");
```

**Comment by elliefm** [comment]
Good catch!  The syslog stuff is pretty new, and looks like not all of it properly ensures it has the syslog replacement before trying to assert a particular string was or wasn't syslogged (since, on all our systems, it always has it) 😅

Probably need to audit all the tests that call getsyslog() and make sure they first check whether they have have_syslog_replacement -- just opened #87 for that
